### ISSUE TITLE
Improve raceWith

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz-zio#361]
Currently, `raceWith` uses `supervised` on left / right:

```scala
    (for {
      done  <- Promise.make[E2, C]
      race  <- Ref[IO.Race](IO.Race.Started)
      left  <- self.fork
      right <- that.fork
      _     <- left.observe.flatMap(arbiter(leftWins, left, right, race, done)).fork
      _     <- right.observe.flatMap(arbiter(rightWins, right, left, race, done)).fork
      c     <- done.get
    } yield c).supervised
```

This means that the `left` and `right` fibers cannot outlive the invocation of either `leftWins` or `rightWins`.

The `supervised` combinator is used to ensure that if the main fiber is interrupted, the `left` and `right` will be interrupted, too. However, we can do that in other ways:

```scala
    for {
      done   <- Promise.make[E2, C]
      race   <- Ref[IO.Race](IO.Race.Started)
      fibers <- Ref[Fiber[E2, Unit]](Fiber.point(()))
      c      <- (for {
                  left   <- self.fork.flatMap(f => fibers.update(_.zip(f).void)).uninterruptibly
                  right  <- that.fork.flatMap(f => fibers.update(_.zip(f).void)).uninterruptibly
                  _      <- left.observe.flatMap(arbiter(leftWins, left, right, race, done)).fork
                  _      <- right.observe.flatMap(arbiter(rightWins, right, left, race, done)).fork
                  c      <- done.get      
                } yield c).onTermination(fibers.get.flatMap(_.interrupt))
    } yield c
```

**Comment by regiskuckaertz** [comment]
What do you think of passing `arbiter` as an argument to `raceWith` (or `race0`)?

**Comment by jdegoes** [comment]
@regiskuckaertz I'm concerned it's very low-level and there's a potential for misuse (what happens if they never complete the promise?).

I'm thinking we should change it like this:

```scala
leftDone: (IO[Nothing, ExitResult[E1, A]], Fiber[E1, B]) => IO[E2, C],
rightDone: (IO[Nothing, ExitResult[E1, B]], Fiber[E1, A]) => IO[E2, C]
```

Then we call `leftDone` when left exits (**not** when it wins), and we call `rightDone` when right exits (**not** when it wins). I think this allows the user to do all possibilities.

**Comment by jdegoes** [comment]
Fixed by #376
